pkg/platform/provider/baremetal/cluster/update.go

```diff
@@ -136,14 +138,27 @@ func (p *Provider) EnsurePreClusterUpgradeHook(ctx context.Context, c *v1.Cluste
 }
 
 func (p *Provider) EnsureUpgradeControlPlaneNode(ctx context.Context, c *v1.Cluster) error {
+	// check all machines are upgraded before upgrade cluster
+	machines, err := p.platformClient.Machines().List(context.TODO(), metav1.ListOptions{
+		LabelSelector: fields.OneTermEqualSelector(constants.LabelNodeNeedUpgrade, "").String(),
+		FieldSelector: fields.OneTermEqualSelector(platformv1.MachineClusterField, c.Name).String(),
+	})
+	if err != nil {
+		return err
+	}
+	if len(machines.Items) != 0 {
+		return fmt.Errorf("some machines, %s..., need to be upgraded", machines.Items[0].Name)
+	}
```

Comment:
1. should return ` errors.Wrap(err, xxx)` to trace the origin error stack
2. machines.Items[0]  VS some machines?